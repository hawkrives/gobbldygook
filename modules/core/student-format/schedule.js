import isString from 'lodash/isString'
import uuid from 'uuid/v4'

import { randomChar } from 'modules/lib'

export function Schedule(data={}) {
	const baseSchedule = {
		id: uuid(),
		active: false,

		index: 1,
		title: `Schedule ${randomChar().toUpperCase()}`,

		clbids: [],
		year: 0,
		semester: 0,

		metadata: {},
	}

	let schedule = {
		...baseSchedule,
		...data,
	}

	if (!isString(schedule.id)) {
		throw new TypeError('Schedule id must be a string.')
	}

	if (typeof schedule.year === 'string') {
		schedule.year = parseInt(schedule.year, 10)
	}

	if (typeof schedule.semester === 'string') {
		schedule.semester = parseInt(schedule.semester, 10)
	}

	return schedule
}
