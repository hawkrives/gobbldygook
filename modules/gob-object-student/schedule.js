import isString from 'lodash/isString'
import uuid from 'uuid/v4'
import {randomChar} from '@gob/lib'

export function Schedule(data = {}) {
	const baseSchedule = {
		id: uuid(),
		active: false,

		index: 1,
		title: `Schedule ${randomChar().toUpperCase()}`,

		clbids: [],
		year: 0,
		semester: 0,

		metadata: {},
	}

	let schedule = {...baseSchedule, ...data}

	if (!isString(schedule.id)) {
		throw new TypeError('Schedule id must be a string.')
	}

	if (schedule.clbids.some(id => typeof id === 'number')) {
		schedule.clbids = schedule.clbids.map(id => typeof id === 'number' ? String(id).padStart(10, '0') : id)
	}

	if (typeof schedule.year === 'string') {
		schedule.year = parseInt(schedule.year, 10)
	}

	if (typeof schedule.semester === 'string') {
		schedule.semester = parseInt(schedule.semester, 10)
	}

	return schedule
}
