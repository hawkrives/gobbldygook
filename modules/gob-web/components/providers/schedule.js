// @flow

import * as React from 'react'
import {List, Map} from 'immutable'
import {countCredits} from '@gob/examine-student'
import {
	Student,
	Schedule,
	type WarningType,
	type FabricationType,
} from '@gob/object-student'
import type {Course as CourseType, CourseError} from '@gob/types'
import {getOnlyCourse, getCourse} from '../../helpers/get-courses'
import {loadDataForTerm} from '../../workers/load-data'

export type ChildrenProps = {
	loading: boolean,
	checking: boolean,
	courses: List<CourseType | FabricationType | CourseError>,
	warnings: Map<string, List<WarningType>>,
	hasConflict: boolean,
	credits: number,
}

type Props = {
	schedule: Schedule,
	student: Student,
	children: ChildrenProps => React.Node,
}

type State = {
	loading: boolean,
	checking: boolean,
	courses: List<CourseType | FabricationType | CourseError>,
	warnings: Map<string, List<WarningType>>,
	hasConflict: boolean,
	credits: number,
}

export class ScheduleProvider extends React.Component<Props, State> {
	state = {
		loading: true,
		checking: true,
		courses: List(),
		warnings: Map(),
		hasConflict: false,
		credits: 0,
	}

	componentDidMount() {
		this.ensureDataExists()
		this.prepare(this.props)
	}

	componentDidUpdate(prevProps: Props) {
		if (this.props.schedule !== prevProps.schedule) {
			this.prepare(this.props)
		}
	}

	ensureDataExists = async () => {
		await loadDataForTerm(parseInt(this.props.schedule.getTerm()))
	}

	prepare = async (props: Props) => {
		let {fabrications} = props.student
		let {schedule} = props

		this.setState(() => ({loading: true, checking: true}))

		let courses = await schedule.getCourses(getCourse, fabrications)
		let credits = countCredits([...courses])

		this.setState(() => ({courses, credits, loading: false}))

		let {warnings, hasConflict} = await schedule.validate(getOnlyCourse)

		this.setState(() => ({warnings, hasConflict, checking: false}))
	}

	render() {
		let {
			courses,
			credits,
			warnings,
			hasConflict,
			loading,
			checking,
		} = this.state

		return this.props.children({
			loading,
			checking,
			courses,
			warnings,
			hasConflict,
			credits,
		})
	}
}
