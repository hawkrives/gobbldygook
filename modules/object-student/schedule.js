'use strict'
const isString = require('lodash/isString')
const uuid = require('uuid/v4')

const { randomChar } = require('../lib/random-char')

function Schedule(data = {}) {
    const baseSchedule = {
        id: uuid(),
        active: false,

        index: 1,
        title: `Schedule ${randomChar().toUpperCase()}`,

        clbids: [],
        year: 0,
        semester: 0,

        metadata: {},
    }

    let schedule = Object.assign({}, baseSchedule, data)

    if (!isString(schedule.id)) {
        throw new TypeError('Schedule id must be a string.')
    }

    if (typeof schedule.year === 'string') {
        schedule.year = parseInt(schedule.year, 10)
    }

    if (typeof schedule.semester === 'string') {
        schedule.semester = parseInt(schedule.semester, 10)
    }

    return schedule
}

module.exports.Schedule = Schedule
