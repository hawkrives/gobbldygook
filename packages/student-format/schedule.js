import isString from 'lodash/isString'
import {v4 as uuid} from 'uuid'

import randomChar from '../helpers/random-char'

export default function Schedule(data={}) {
	const baseSchedule = {
		id: uuid(),
		active: false,

		index: 1,
		title: `Schedule ${randomChar().toUpperCase()}`,

		clbids: [],
		year: 0,
		semester: 0,

		metadata: {},
	}

	let schedule = {
		...baseSchedule,
		...data,
	}

	if (!isString(schedule.id)) {
		throw new TypeError('Schedule id must be a string.')
	}

	if (typeof schedule.year === 'string') {
		schedule.year = parseInt(schedule.year, 10)
	}

	if (typeof schedule.semester === 'string') {
		schedule.semester = parseInt(schedule.semester, 10)
	}

	return schedule
}
