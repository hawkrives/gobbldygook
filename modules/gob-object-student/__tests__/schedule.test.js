// @flow

import stringify from 'stabilize'
import {Schedule} from '../schedule'
import {List} from 'immutable'

describe('Schedule', () => {
	it('does not mutate the passed-in object', () => {
		const clbids = []
		const input = {clbids}
		new Schedule(input)
		expect(input.clbids).toBe(clbids)
	})

	it('copies from one Schedule to another', () => {
		let initial = new Schedule({
			id: 'null',
			active: true,
			index: 10,
			title: 'title me oh my',
			clbids: ['1'],
			year: 2018,
			semester: 59,
			metadata: {key: 'value'},
		})
		let copy = new Schedule(initial)

		expect(copy.id).toBe(initial.id)
		expect(copy.active).toBe(initial.active)
		expect(copy.index).toBe(initial.index)
		expect(copy.title).toBe(initial.title)
		expect(copy.clbids).toBe(initial.clbids)
		expect(copy.year).toBe(initial.year)
		expect(copy.semester).toBe(initial.semester)
		expect(copy.metadata).toBe(initial.metadata)
	})

	it('uses the ID that you give it', () => {
		let schedule = new Schedule({id: '1'})
		expect(schedule.id).toBe('1')
	})

	it('creates a unique ID for each new schedule without an ID prop', () => {
		let sched1 = new Schedule()
		let sched2 = new Schedule()
		expect(sched1.id).not.toBe(sched2.id)
	})

	it('holds a schedule for a student', () => {
		let sched = new Schedule({
			id: '1',
			active: true,
			year: 1994,
			semester: 3,
			index: 2,
			title: 'My Schedule',
			clbids: ['123', '234', '345'],
		})
		expect(sched).toMatchInlineSnapshot(`
Immutable.Record {
  "id": "1",
  "active": true,
  "index": 2,
  "title": "My Schedule",
  "clbids": Immutable.List [
    "123",
    "234",
    "345",
  ],
  "year": 1994,
  "semester": 3,
  "metadata": Immutable.Map {},
}
`)
	})

	it('can turn into JSON', () => {
		let result = stringify(new Schedule({id: '1', title: 'Schedule 6'}))
		expect(result).toMatchInlineSnapshot(
			`"{\\"active\\":false,\\"clbids\\":[],\\"id\\":\\"1\\",\\"index\\":1,\\"metadata\\":{},\\"semester\\":0,\\"title\\":\\"Schedule 6\\",\\"year\\":0}"`,
		)
	})

	it('converts numeric clbids to strings', () => {
		let sched = new Schedule({
			clbids: [123, 234, 345],
		})

		expect(sched.clbids).toEqual(
			List(['0000000123', '0000000234', '0000000345']),
		)
	})
})
