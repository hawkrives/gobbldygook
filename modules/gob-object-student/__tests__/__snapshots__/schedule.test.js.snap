// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule can turn into JSON 1`] = `"{\\"active\\":false,\\"clbids\\":[],\\"id\\":\\"1\\",\\"index\\":1,\\"metadata\\":{},\\"semester\\":0,\\"title\\":\\"Schedule 6\\",\\"year\\":0}"`;

exports[`Schedule holds a schedule for a student 1`] = `
Object {
  "active": true,
  "clbids": Array [
    "0000000123",
    "0000000234",
    "0000000345",
  ],
  "id": "1",
  "index": 2,
  "metadata": Object {},
  "semester": 3,
  "title": "My Schedule",
  "year": 1994,
}
`;
